### PR TITLE
Clean up masking logic

### LINKED ISSUE #415: Support BitMask Masks
The core KBMOD search should be redesigned to take masks as bitmaps and apply them during stages like create Psi and Phi. The logic for generating the bit mask from the full mask (a bitvector of reasons), including growing the masks, should be moved to preprocessing (standardizer).

The core search logic will need to be updated to handle inserting NO_DATA into psi curves, phi curves, light curves, and stamps. This could be done by applying the bit mask once on load or during each action.

### PR DESCRIPTION
**Pull request by jeremykubica**
This PR refactors the masking code to clean up the interfaces, but does not change end-to-end behavior. A major part of this change is that most masking operations will now work on the `LayeredImage`'s mask layer. For example, `grow_mask` will now expand the masked pixels in the mask layer instead of the science layer.  `LayeredImage`'s `apply_mask()` function will then apply the masks to the science and variance layers.

The motivations for these changes are:
- Support binary masks coming via the standardizers and `WorkUnit`s. These will be cases where all the masking logic is applied when the `WorkUnit` is constructed.
- Maintain backwards compatibility with current input files.
- Consolidate most masking logic in `LayeredImage` instead of across `RawImage`, `LayeredImage`, and `ImageStack`.
- Reduce logic in `RawImage` for possible eventual removal of that class.

More detail on the changes:

`LayeredImage`: Move the majority of the masking logic into `LayeredImage` and perform most changes on the mask layer (instead of directly onto science and variance).

`RawImage`: Remove `grow_mask` function. All mask changes should be done through `LayeredImage` and then applied to `RawImage` with `apply_mask()`.

`ImageStack`:
- Remove masking functions that just loop over the individual `LayeredImage`s, such as `apply_mask_flags()`.
- Remove global mask as an object attribute of `ImageStack`. Replace it with a function to create a new global mask on-demand. This allows the code to skip allocating a global mask in cases where no global masking is done.

`masking.py`:
- Remove the class-based logic. This was added to support use cases that will be covered in the standardizer. We only need to apply the individual functions from `LayeredImage` and `ImageStack`. The new flow is much simpler.

`run_search.py`:
- Move the masking logic into `masking.py`.

Closes #415 and #352.

### CODE REVIEW

File: src/kbmod/masking.py [hunk 1]
```diff
@@ -98,126 +15,70 @@ class DictionaryMasker(BitVectorMasker):
...
+            stack.get_single_image(i).union_masks(global_binary_mask)
+
+    # Grow the masks.
+    if config["mask_grow"] and config["mask_grow"] > 0:
+        for i in range(stack.img_count()):
+            stack.get_single_image(i).grow_mask(config["mask_grow"])
+
+    # Apply the masks to the images.
+    for i in range(stack.img_count()):
+        stack.get_single_image(i).apply_mask(0xFFFFFF)
```
Comment (wilsonbb) [inline src/kbmod/masking.py:82]: Nit: can you add a comment here to note that 0xFFFFFF is being used to apply all flags?

File: src/kbmod/search/layered_image.h [hunk 2]
```diff
@@ -76,7 +93,7 @@ class LayeredImage {
     PSF psf;
     RawImage science;
     RawImage mask;
-    RawImage variance;
+    RawImage variance;
```
Comment (wilsonbb) [inline src/kbmod/search/layered_image.h:96]: Nit: remove trailing whitespace

File: src/kbmod/search/pydocs/image_stack_docs.h [hunk 3]
```diff
@@ -32,57 +32,25 @@ static const auto DOC_ImageStack_build_zeroed_times = R"doc(
   in the stack and the first image.
   ")doc";
 
-static const auto DOC_ImageStack_apply_mask_flags = R"doc(
-  Applies a mask to each image by comparing the given bit vector with the
-  values in the mask layer and marking pixels NO_DATA. Modifies the image in-place.
+static const auto DOC_ImageStack_make_global_mask = R"doc(
+  Create a new global mask that from a set of flags and a threshold.
```
Comment (wilsonbb) [inline src/kbmod/search/pydocs/image_stack_docs.h:36]: Nit: s/global mask that from/global mask from

File: tests/test_layered_image.py [hunk 4]
```diff
@@ -126,36 +145,78 @@ def test_overwrite_PSF(self):
...
         self.assertLess(science_pixel_psf1, science_pixel_psf2)
 
-    def test_mask_threshold(self):
+    def test_binarize_mask(self):
+        # Mask out a range of pixels.
+        mask = self.image.get_mask()
+        for x in range(9):
+            mask.set_pixel(10, x, x)
+
+        # Only keep the mask at for pixels with flags at
```
Comment (wilsonbb) [inline tests/test_layered_image.py:154]: Nit: s/mask at for/mask for

File: tests/test_layered_image.py [hunk 5]
```diff
@@ -126,36 +145,78 @@ def test_overwrite_PSF(self):
...
                 if value > threshold:
                     index = self.image.get_width() * y + x
                     masked_pixels[index] = True
-
-        # Do the masking and confirm we have masked
-        # at least 1 pixel.
-        self.image.apply_mask_threshold(threshold)
         self.assertGreater(len(masked_pixels), 0)
 
+        # Reset the mask an perform threshold masking.
```
Comment (wilsonbb) [inline tests/test_layered_image.py:207]: Nit: s/mask an perform/mask and perform

**Review by wilsonbb — APPROVED** [review]
LGTM, just a few nits

File: src/kbmod/masking.py (hunk 1 above)
Comment (jeremykubica) [reply src/kbmod/masking.py:82]: Added

File: src/kbmod/search/layered_image.h (hunk 2 above)
Comment (jeremykubica) [reply src/kbmod/search/layered_image.h]: Removed

File: src/kbmod/search/pydocs/image_stack_docs.h (hunk 3 above)
Comment (jeremykubica) [reply src/kbmod/search/pydocs/image_stack_docs.h]: Fixed

File: tests/test_layered_image.py (hunk 4 above)
Comment (jeremykubica) [reply tests/test_layered_image.py]: Fixed

File: tests/test_layered_image.py (hunk 5 above)
Comment (jeremykubica) [reply tests/test_layered_image.py]: Fixed

**Pull request merged**